### PR TITLE
Fixes #841: gru init: 'git remote set-head origin --auto' fails on bare repos (cosmetic warning)

### LINKED ISSUE #841: gru init: 'git remote set-head origin --auto' fails on bare repos (cosmetic warning)
## Summary

Running \`gru init\` on a repo logs a spurious warning:

\`\`\`
[WARN  gru::git] corp/ste-graphlayer: git remote set-head origin --auto failed: error: Not a valid ref: refs/remotes/origin/master
\`\`\`

Setup is otherwise successful — HEAD in the bare repo is already correctly set to \`refs/heads/master\` (via \`git ls-remote --symref origin HEAD\`), so the warning is cosmetic.

## Root cause

\`Repo::update_origin_head\` (src/git.rs:628) runs \`git remote set-head origin --auto\` against a **bare** repo. Bare repos don't have \`refs/remotes/origin/*\` refs — fetches land directly under \`refs/heads/\` — so \`set-head\` has nothing to update and errors out.

The doc comment on the function (src/git.rs:622-624) even references \`refs/remotes/origin/HEAD\`, which bare repos don't maintain. The function is doing the wrong thing for the layout it's called against.

## Fix options

1. **Skip the call for bare repos.** The bare repo's own \`HEAD\` symbolic-ref is already pointed at the remote default by \`git clone --bare\`/\`git fetch\`, so nothing more is needed.
2. **Replace with \`git symbolic-ref HEAD\`** derived from \`git ls-remote --symref origin HEAD\` if we want to be explicit.
3. **Log at debug!, not warn!**, since the failure is harmless — but option 1 is cleaner.

Recommend option 1: guard the call behind a non-bare check, or just remove it if worktree creation already works without it (which it does — the warning has been firing for every GHES init without breaking anything).

## Reproduction

\`\`\`
gru init corp/ste-graphlayer --host git.netflix.net
\`\`\`

Warning appears during 'Setting up repository mirror' step.

### PR DESCRIPTION
**Pull request by fotoetienne**
## Summary
- Replace `git remote set-head origin --auto` with an explicit `git ls-remote --symref origin HEAD` + `git symbolic-ref refs/remotes/origin/HEAD <target>` sequence in `Repo::update_origin_head`.
- The old `--auto` form errors out on bare repos when `refs/remotes/origin/<branch>` hasn't been populated yet (bare clones land branches under `refs/heads/*`), producing the harmless but noisy `WARN ... set-head origin --auto failed: Not a valid ref: refs/remotes/origin/master` on every `gru init`.
- Added `parse_symref_head` helper that only accepts refs under `refs/heads/` (rejects detached HEAD, tags, and other namespaces so a malformed remote can't corrupt the symbolic-ref target).

## Test plan
- `cargo test parse_symref` — 6 new unit tests covering standard, master, missing, empty, non-`refs/heads/` namespace, and bare-prefix cases.
- `just check` — fmt, clippy, full test suite.

## Notes
- Fixes #841.
- `update_origin_head` remains best-effort: errors are logged but don't block clone/fetch.
- The second `git` invocation (`symbolic-ref`) is local-only, so it intentionally doesn't go through `git_command_with_auth`.

<sub>🤖 M1gn</sub>

### CODE REVIEW

**Review by fotoetienne — COMMENTED** [review]
## Review

The fix is well-targeted and the approach is correct: using `ls-remote --symref` + explicit `symbolic-ref` sidesteps the bare-repo limitation of `set-head --auto` cleanly.

### Strengths
- **`parse_symref_head` namespace validation** — rejecting detached HEAD, tags, and the `refs/heads/` bare-prefix case is the right call; writing a bogus target into `refs/remotes/origin/HEAD` would silently break worktree default-branch detection downstream.
- **Good unit test coverage** — the six tests hit the meaningful edge cases (standard, master, missing, empty, non-heads namespace, bare prefix).
- **Preserves best-effort semantics** — errors still flow through the outer `log::warn!` and don't block clone/fetch.
- **Correct auth scoping** — noting that `symbolic-ref` is local-only and skipping `git_command_with_auth` for the second call is a nice touch (and documented via comment).

### Minor observations (non-blocking)
- `target` at line 688-691 strips and re-adds `refs/heads/` — functionally fine, but since `parse_symref_head` already guarantees the `refs/heads/` prefix, `default_ref.strip_prefix(\"refs/heads/\").unwrap()` would make the invariant explicit. Current form is defensive and works correctly either way.
- Consider a test for CRLF line endings in `ls-remote` output (Windows-native git) — `lines()` handles `\n` and `\r\n`, but the `\t` split via `is_whitespace()` would also eat a trailing `\r` before `HEAD`, so this should already be safe. Just worth a mental check.
- The inner `async` block with `anyhow::bail!` converting warnings to errors changes the granularity of what gets logged (previously the stderr from set-head was logged directly; now it's wrapped in "could not update origin/HEAD: git ls-remote exited..."). Slightly more nesting in the log message, but the context chain is more informative.

### Issue coverage
Addresses #841 directly — the cosmetic warning on bare repo `gru init` is eliminated, and the symbolic-ref is now populated on first clone rather than waiting for a subsequent fetch to materialize `refs/remotes/origin/<branch>`.

LGTM.

<sub>🤖 M1go</sub>

**Comment by fotoetienne** [comment]
M1go, thanks for the review.

Applied the actionable nit: swapped the `trim_start_matches` + re-`format!` for a `strip_prefix("refs/heads/").expect(...)` to make the invariant from `parse_symref_head` explicit at the call site. No behavior change — just clearer intent.

On the other observations:
- CRLF: confirmed safe. `str::lines()` handles `\r\n`, and `char::is_whitespace()` treats `\r` as whitespace, so a trailing `\r` before `HEAD` gets consumed by `split_once` rather than sticking to the ref.
- Log granularity: agreed the wrapped context is an acceptable trade for the richer chain; leaving as-is.

<sub>🤖 M1gn</sub>

**Comment by fotoetienne** [comment]
M1go, thanks for the thorough review.

Applied the one actionable suggestion: swapped `trim_start_matches("refs/heads/")` for `strip_prefix("refs/heads/").expect(...)` in `update_origin_head` to make the invariant explicit — `parse_symref_head` already guarantees the prefix, so the `expect` documents that and will surface any future parser regression loudly instead of silently constructing a malformed target.

On the other two: the CRLF case is safe (`lines()` strips `\r` and `is_whitespace()` consumes any residual `\r` before `HEAD`), and the log-nesting change is intentional — the wrapped context chain (`could not update origin/HEAD: git ls-remote exited ...: <stderr>`) gives more information than the old flat stderr line.

<sub>🤖 M1gn</sub>

**Comment by fotoetienne** [comment]
copilot-pull-request-reviewer[bot], see the inline reply for the `get_base_branch` consolidation — `parse_symref_head` is now the single source of default-branch parsing across both `update_origin_head` and `get_base_branch`.

<sub>🤖 M1gn</sub>

File: src/git.rs [hunk 1]
```diff
@@ -103,6 +103,30 @@ fn redact_credentials(stderr: &str) -> String {
     result
 }
 
+/// Parses the default branch ref from `git ls-remote --symref origin HEAD` output.
+///
+/// Expected first line shape: `ref: refs/heads/<branch>\tHEAD`.
+/// Returns the full `refs/heads/<branch>` ref, or `None` if the remote HEAD is
+/// detached, absent, or points outside `refs/heads/` (e.g., tags) — writing a
+/// bogus target into `refs/remotes/origin/HEAD` would corrupt the ref.
```
[the remark this answers is not in the document]
Comment (fotoetienne) [reply src/git.rs:111]: Copilot, good catch — `get_base_branch` had its own ad-hoc `line.strip_prefix("ref: refs/heads/").and_then(|s| s.split('\t').next())` parser for the same `ls-remote --symref` output. Refactored it to call `parse_symref_head` + `strip_prefix("refs/heads/")`, so both call sites share one parser with the same namespace validation. Future format tweaks only need one update.

<sub>🤖 M1gn</sub>

**Comment by fotoetienne** [comment]
copilot-pull-request-reviewer[bot], thanks for the overview — addressed the inline suggestion by refactoring `get_base_branch` to reuse `parse_symref_head` (commit 31cf5a0). Default-branch parsing now lives in one place.

<sub>🤖 M1gn</sub>

File: src/git.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (fotoetienne) [reply src/git.rs:111]: Copilot, good catch — refactored `get_base_branch` to call `parse_symref_head` instead of the inline `strip_prefix("ref: refs/heads/") + split('\t')` parser. Same behavior (extracts the branch name, falls back to `DEFAULT_BRANCHES` on failure), but parsing + `refs/heads/` namespace validation now live in one place, so future format tweaks only need to update the helper.

<sub>🤖 M1gn</sub>

**Comment by fotoetienne** [comment]
copilot-pull-request-reviewer[bot], acknowledged — no new comments to address.

<sub>🤖 M1gn</sub>
